source/common/http/http1/codec_impl.h

```diff
@@ -367,39 +353,42 @@ class ConnectionImpl : public virtual Connection,
    * Called when a request/response is beginning. A base routine happens first then a virtual
    * dispatch is invoked.
    */
-  virtual Status onMessageBegin() PURE;
+  int onMessageBegin() override;
 
   /**
    * Called when URL data is received.
    * @param data supplies the start address.
    * @param length supplies the length.
    */
-  virtual Status onUrl(const char* data, size_t length) PURE;
+  int onUrl(const char* data, size_t length) override;
+  virtual Status onUrlStatus(const char* data, size_t length) PURE;
 
   /**
    * Called when header field data is received.
    * @param data supplies the start address.
    * @param length supplies the length.
    * @return A status representing success.
    */
-  Status onHeaderField(const char* data, size_t length);
+  int onHeaderField(const char* data, size_t length) override;
+  Status onHeaderFieldStatus(const char* data, size_t length);
```

Comment:
these
```cpp
int foo(const char* data, size_t length)
Status fooStatus(const char* data, size_t length)
```
methods were a bit of a quick hack for me to work around the newer `RETURN_IF_ERROR`/Status stuff, but I think we can do better. Since the callback functions return ints, we basically need a way to call `setAndCheckCallbackStatus` on the result of `RETURN_IF_ERROR` and return _that_, without which we get a type error trying to return a Status.

Reply:
Yep I agree, we essentially have extra functions defined. 

Ideally I think that the `int` parser-dependent status code should be localized to the parser code and codec_impl.cc shouldn't see them.

I wonder if I can do a `parser_->statusToInt(x)` in the parser code, but make `ParserStatus setAndCheckCallbackStatus` run at the end of `foo`?

Reply:
Best I can come with at the moment is maybe making `codec_status_` a part of the parser library rather than the codec library. Reading over the codec it _seems_ to primarily be more informational about the parser library, and might make more sense there with the codec calling something like `parser_->status().ok()`. That would enable to do `setAndCheckCallbackStatus` in the parser and have the `ParserCallbacks` methods just return a `Status` (thus the `RETURN_IF_ERROR` "just works"), keeping a lot of that flow the same as what we currently do with the http_parser hardcoded in.

So rather than teaching the codec to return `int`s for the `ParserCallback`, we teach the `ParserCallback` to accept `Status` and do the `setAndCheckCallbackStatus` conversion.

e.g. pre-PR
```cpp
// codec_impl.cc
    [](http_parser* parser) -> int {
      auto* conn_impl = static_cast<ConnectionImpl*>(parser->data);
      auto status = conn_impl->onMessageBeginBase();
      return conn_impl->setAndCheckCallbackStatus(std::move(status));
    },
```
would become
```cpp
// legacy_parser_impl.cc
        [](http_parser* parser) -> int {
          auto* conn_impl = static_cast<ParserCallback*>(parser->data);
          auto status = conn_impl->onMessageBegin();
          return conn_impl->setAndCheckCallbackStatus(std::move(status));
        },
```
It'd be a bigger mechanical change though

Reply:
Hmmm yeah! That works! The codec can call into the parser for the status. Will implement.

Reply:
YAY! all done.